Todak 79 [feat] 일기 작성 API 개발

kchsoft opened the pull request:
<details>
<summary> PR 체크리스트</summary>
  
- [x] 🔀 PR 제목의 형식을 잘 작성했나요? e.g. #1 [feat] 소셜 로그인 구현
- [x] 💭 이슈는 등록했나요?
- [x] 🏷️ 라벨은 등록했나요?
- [x] 🙇‍♂️ 리뷰어를 지정했나요? (페어가 존재할 시)
</details>

## 변경사항

### Feature
- ~일기 관련 객체를 생성~
- ~일기 작성 API 컨트롤러/서비스/레포지토리를 생성 (컨트롤러 경우 임시로 인증 객체를 받아옴.)~
- ~일기 작성 API는  AI 서버에게 비동기로 요청(WebFlux).~
- ~일기 작성 API Swagger 적용~
---
2024/10/21
- ai 코멘트 즉시 응답 기능 추가
- 일기 작성 초과 ```Exception``` 추가
- ERD 데이터 타입 및 제약 조건을  entity 필드에 적용
- Diary의 publicContent 내용 분리 및 Visibility 내용 삭제

### Test
- ~SpringBoot 통합 테스트에 사용할  Base 객체 생성~
- ```AiService Test``` 의 목적을 __"비동기로 요청을 보내는지 테스트"__ -> __"요청이 정상적으로 전달 되는지"__ 로 변경하였습니다! webclient를 처음 적용했기에 해당 코드가 비동기 요청을 보내는지 확인이 필요했는데요! 사실상 비동기 요청이 나가는 것을 확인 했다면, webflux 라이브러리가 바뀌지 않는 이상 요청은 비동기로 나갈 것 이고, 더 중요한 것은 요청이 제대로 가는지 중요하다고 생각되어 "요청 정상 전달"을 목적으로 두어 테스트를 바꾸었습니다!
- ```DiaryRepository``` 테스트 추가

### Config
- Test Config의 H2 관련 설정 삭제
- AI 서버 Test url 추가
- entity 데이터 타입 및 제약 변경 사용 적용을 위해 __jpa의 auto-ddl을 create로 설정__

## 리뷰 받고 싶은 내용
- ~비동기 요청을 테스트 하는 방법을 고민하다, 함수 call 순서와, mockServer에 도달하는 순서가 다를 때 까지 반복하는 테스트를 만들었습니다. 시간을 고려한 비동기 테스트도 고려했지만, 복잡한 것 같아 함수 call 순서에 따른 테스트를 만들었는데 피드백 있으시다면 말씀해 주세요 :)~ 
- ~코드가 계속 길어져 지금이라도 PR 요청 보냅니다.~
---


## 테스트 및 결과
- local에서 모든  테스트 통과합니다.

## 다음 할 일
- ~AI 위로 코멘트는 응답을 받아 사용자에게 전달로 변경~
- ~AiServcie 테스트 수정~
- ~하루 일기 작성량 초과시 에러 반환~
- ~Audit 적용~ -> 인증 객체 개발 완료시 Audit 적용 예정
- ~SpringBoot 통합 테스트 extend 적용~

close #10

kchsoft committed:
feat: Diary 관련 객체 생성 TODAK-79

kchsoft committed:
feat: BaseEntity 생성 및 적용 TODAK-79

- DiaryEntity에 Base 적용

kchsoft committed:
feat: 일기 작성 API 개발 완료 TODAK-79

kchsoft committed:
test: 일기 작성 API 비동기 통신 확인 TODAK-79

- ai 서버 설정 분리
- 테스트 위한 mock server 의존성 추가

kchsoft committed:
test: springboot 테스트 환경 설정 TODAK-79

- springboot test에 사용할 test yaml 파일 활성화 적용
- springboot test 설정 추상화 적용

kchsoft committed:
refactor: 불필요한 에노테이션 제거 TODAK-79

kchsoft committed:
refactor: 시간 객체 변경 TODAK-79

kchsoft committed:
refactor: ai와 diary 서비스 코드 분리 TODAK-79

kchsoft committed:
feat: 일기 작성 controller 개발 TODAK-79

kchsoft committed:
fix: stash 오류로 인한 파일 복구 TODAK-79

kchsoft committed:
refactor: DB 시간 데이터 타입 변경 TODAK-79

kchsoft committed:
Merge branch 'develop' into TODAK-79

kchsoft committed:
Merge branch 'develop' into TODAK-79

kchsoft committed:
feat: 일기 작성 API Swagger 적용 TODAK-79

kchsoft committed:
config: 불필요한 설정 및 필요한 설정 추가 TODAK-79

- h2 관련 Datasource 삭제
- ai server 테스트 url 추가

Urchinode commented on src/main/java/com/heartsave/todaktodak_api/ai/dto/AiContentRequest.java:
```diff
@@ -0,0 +1,13 @@
+
+@Builder
+@Getter
+public class AiContentRequest { // springboot request to ai server
```
머지 후에 `characterInfo`가 추가되는게 맞을까요?

kchsoft replied:
아! 추가하는걸 깜박했습니다! 감사합니다!
현재 ```MemberEntity```에 ```characterInfo``` 가 ```Obejct```인데 ```String```으로 바꿀까요??

Urchinode replied:
아아, 그러네요..
String으로 하자니, 머지되면 타입이 어긋날 것이고
Object로 하자니 머지해도 다시 수정해야겠군요 😂

그러면 지금은 추가하지 않고, 머지 후 새로운 브랜치에 추가하는 방식으로 할까요?

kchsoft replied:
아하 네! 좋습니당!

Urchinode commented on src/main/java/com/heartsave/todaktodak_api/ai/service/AiService.java:
```diff
@@ -0,0 +1,67 @@
+
+  private final WebClient webClient;
+
+  public void callAiContent(DiaryEntity diary) {
```
추후 각 call 메서드에 대한 결과를 응답받아 컨트롤러에게 전달되는건가요?
이를 테면,
```java
String comment = callComment(request);
// ...(중략)
return ContentResponseDto(comment, arg1, arg2, ...);
```

이런 식으로 리팩토링되는건가요?

kchsoft replied:
네 맞습니다! AiContentResponse 객체를 만들어 Controller에서 반환할 예정입니다!

Urchinode commented on src/main/java/com/heartsave/todaktodak_api/diary/common/DiaryContentConstraintConstant.java:
저도 제가 만든 dto의 validation에 대해 상수화해서 관리해야겠네요, 👍👍

Urchinode commented on src/main/java/com/heartsave/todaktodak_api/diary/constant/DiaryEmotion.java:
`diary/common`와 중복되는 클래스인지 궁금합니다

kchsoft replied:
아! 오! 저번에 stash를 잘 못 적용해서 conflict이 발생했고, conflict를 잘 못 해결했는데요...! 그래서 이런 결과가 나온 것 같습니다!ㅠㅠ 저도 의문인 부분이 있었는데 알려주셔 감사합니다.👍 아마 constant 패키지만 남기고 common을 우선 삭제할 것 같습니다!

Urchinode commented on src/main/java/com/heartsave/todaktodak_api/diary/controller/DiaryController.java:
```diff
@@ -0,0 +1,44 @@
+      })
+  @PostMapping("/my")
+  public ResponseEntity<Void> writeDiary(
+      @AuthenticationPrincipal OAuth2User auth,
```
제가 작업하고 있는 브랜치에서 인증 정보 구현체(`TodakUser`)를 만들고 있는데,
merge후 반영하도록 하겠습니다 👍

kchsoft replied:
넵 감사합니다!

Urchinode commented on src/main/java/com/heartsave/todaktodak_api/diary/entity/DiaryEntity.java:
```diff
@@ -0,0 +1,81 @@
+    initialValue = 1,
+    allocationSize = 50)
+@Table(name = "diary")
+public class DiaryEntity extends BaseEntity {
```
저희 erd 설계 작업으로 결정된 varchar 길이에 대해 `@Column`에 `length`도 설정하는건 어떨까요?

kchsoft replied:
네 좋습니다!
emotion과 url 2개 부분에 적용할게요!

Urchinode commented on src/test/java/com/heartsave/todaktodak_api/config/BaseSpringbootTest.java:
이 클래스는 추후 만들어질 테스트 코드에 사용되는 클래스인가요?
`TodaktodakApiApplicationTests`와 내용이 같아서 궁금했습니다

kchsoft replied:
네! springboot 통합 테스트 클래스를 만들 때 마다 ```@ActiveProfiles("test")``` 를 붙이는게 번거로울 것 같기도 하고, 이후에 추가되는 에노테이션이 있다면 한 번에 적용하기 위해서 만들었습니다!
```TodaktodakApiApplicationTests``` 에 있는 에노테이션은 지우고 Base를 extends하여 반영하겠습니다!

Urchinode commented on src/test/java/com/heartsave/todaktodak_api/ai/service/AiServiceTest.java:
```diff
@@ -0,0 +1,97 @@
+      request = mockWebServer.takeRequest();
+      requestOrder.add(request.getPath());
+    }
+    assertThat(methodCallOrder.equals(requestOrder)).isFalse();
```
요청 순서와 메서드 실행 순서가 다름을 통해 비동기 요청인 것을 판단하는게 맞나요?
세 개의 메서드만 존재하다보니 요청과 실행 순서가 다른 경우를 얻기까지의 시간이 오래 걸리는 편인가요?

kchsoft replied:
네 맞습니다!
대략적인 경험상 테스트 10번을 돌리면 1번 정도는 호출 순서와 동일한 순서가 나와, 테스트를 한 번 더 반복하였습니다!

Urchinode replied:
오호, while 루프를 10번 내로 진행하면 requestOrder와 methodCallOrder의 순서가 다른 경우가 만들어진다는 말씀인가요?
생각보다 아주 빨리 발생하네요 👍👍

kchsoft replied:
네! 테스트를 돌리면 90%확률로 한 번에 통과하고, 10% 확률로 while문이 한 번 더 돌아간다고 생각하시면 될 것 같아요!

이전에 달아주신 MockResponse관련 해서도 한 번 알아보고 있습니다! 정보 공유해 주셔서 감사해요~!!👍

Urchinode replied:
아하, 좋습니다~
merge 편하신때 해주시면 감사하겠습니다 ^0^

Urchinode reviewed (commented):
제가 비동기 요청 확인에 대한 테스트를 진행한 적이 없어서 방법을 모르는데,
에반님의 mock server 테스트로 어떻게 테스트해야하는지 알수 있어서 좋았습니다 👍

mockwebserver를 사용한적이 없어서, 잘모르겠지만
요청 순서가 다를때까지 진행되야 하는 현재 테스트(`while` 조건문)를 다음과 같이 변경하면
while문 없이 비동기 요청 여부를 바로 확인해볼 수 있지않을까 생각됩니다.

검색하니 `MockResponse`에 http 헤더나 본문을 delay하는 메서드가 있는것 같은데, 이걸로 개선이 될까요?
(`setHeadersDelay`, `setBodyDelay`)

https://www.programcreek.com/java-api-examples/?class=okhttp3.mockwebserver.MockResponse&method=setBodyDelay

kchsoft committed:
feat: 일기 작성 API 보충 개발 TODAK-79

- ai 위로 코멘트는 즉시 응답으로 변경
- 하루 일기 작성 제한(1일 1개)
- Diary 관련 이름 변경(post -> diary)
- 하루 일기 초과량 에러 추가

kchsoft committed:
test: springboot 테스트 에노테이션 상속화 TODAK-79

- Base 테스트 객체를 상속받아 , 불필요한 에노테이션 제거

kchsoft committed:
chore: 중복된 상수/enum 객체 제거 TODAK-79

kchsoft committed:
test: ai 컨텐츠 요청 테스트 수정 TODAK-79

- ai 코멘트 응답 확인
- 테스트 메서드 정의 변경(비동기 체크 -> ai 요청 체크)

kchsoft committed:
refactor: DiaryEntity 데이터 타입 변경 TODAK-79

- ERD 바탕으로 데이터 타입,길이 제한 적용

kchsoft committed:
config: ai 서버 url 추가 변경 TODAK-79

kchsoft committed:
feat: 일기 작성시 ai 코멘트 추가 TODAK-79

kchsoft committed:
refactor: BaseEntity 필드 데이터 타입 변경 TODAK-79

- 데이터 생성/수정 구분자 타입 Long 으로 변경

kchsoft committed:
feat: 공개 일기 내용 분리 TODAK-79

- 기획 변경으로 인해 Public Diary RED를 만들게 되어, Public 관련 내용 삭제

kchsoft committed:
feat: Diary 서비스 트랙잭션 적용 TODAK-79

kchsoft committed:
style: google 포맷팅 적용 TODAK-79

kchsoft committed:
test: diary 서비스 테스트 TODAK-79

- 일기 생성, 하루 일기 생성 횟수 초과 에러 테스트 작성
- 서비스 계층 로직 오류 해결

JaeJunday commented:
CI 과정을 일부 업데이트했는데 반영되지 않은것같습니다. develop을 머지한 뒤 다시 푸시부탁드립니다.

kchsoft committed:
Merge branch 'develop' into TODAK-79

kchsoft committed:
test: diary 레포지토리 테스트 TODAK-79

- 날짜에 해당하는 일기 찾기 테스트
- 테스트에 필요한 기본 객체(member, diary) 생성 클래스 추가 및 적용
- 레포지토리 메서드에 종속적이지 않고, 범용적인 jpql 적용

Urchinode reviewed (approved):
`BaseTestObject`로 테스트 데이터를 관리하는게 좋은 것 같습니다👍
